Gloss CSS properties, values, units, datatype

cvrebert opened the pull request.

zcorpan commented:
If we want to have this consistently done, it seems better to me to implement this in wattsi, similar to what bikeshed does. But maybe not, I'm not sure!

```
$ grep -Eo "'[a-z-]+'" source | sort | uniq -d
'a'
'alphabetic'
'anchor-point'
'auto'
'background-color'
'background-image'
'binding'
'black'
'block'
'blue'
'border-bottom-width'
'border-left-width'
'border-right-width'
'border-top-width'
'bottom'
'canvas'
'center'
'checkbox'
'clear'
'click'
'color'
'content'
'critical-no-alt'
'cursor'
'data'
'data-hp'
'direction'
'display'
'file'
'float'
'font'
'font-size'
'font-weight'
'green'
'height'
'hello'
'iframe'
'inherit'
'inline-block'
'input'
'justify'
'left'
'letter-spacing'
'line-height'
'load'
'logout'
'ltr'
'margin-bottom'
'margin-left'
'margin-right'
'margin-top'
'message'
'middle'
'move'
'name'
'none'
'normal'
'onclick'
'overflow'
'p'
'padding-right'
'playing'
'position'
'pre'
'pre-wrap'
'red'
'right'
'round'
'rtl'
'script'
'scroll'
'search-done'
'search-error'
'search-result'
'silver'
'span'
'src'
'string'
'td'
'test'
'text-align'
'top'
'unicode-bidi'
'vertical-align'
'video'
'white-space'
'width'
'x'
```

domenic commented:
I do think that would be ideal. But it's probably a decent amount of work... we'd have to integrate with all those databases and such...

zcorpan commented:
Yeah. How about we go ahead with this PR, but go over the list above and add any missing properties?

tabatkins commented:
The Shepherd datafiles aren't difficult to process. I'm happy to help guide someone thru them if they're trying to implement something in Wattsi.

domenic commented:
It occurs to me that HTML's source doesn't have any precedent for autolinking without a tag delimiter. That is, we always wrap in at least `<span>`. If we could reduce that to `<a>` I would be extremely happy, and I know we have that planned. But I am not sure that we need an extra Wattsi step just to allow `'color'` instead of `<a>'color'</a>`, especially given the potential false-positives (e.g. I presume `'x'`, `'td'`, and `'src'` from the above list are not supposed to be CSS linked) and the continuing issue of which is the best CSS spec for defining a property (see e.g. https://github.com/tabatkins/bikeshed/issues/622).

In other words, I'd be OK with just finishing up this PR instead of tooling out Wattsi.

domenic commented:
My other concern with tooling out Wattsi is that unlike Bikeshed, Wattsi doesn't have an independent concept of "specs database" that updates independently from either the tool or the input source.

We'd have to either introduce that, or maintain our own database (presumably Shepherd-derived) in the whatwg/html or whatwg/html-build repo that Wattsi uses, or something else entirely.

cvrebert commented:
Added more commits. This now glosses every property I could find, and some (albeit not all) values+units+datatypes.

zcorpan commented:
I'm not sure we should xref property values. Whenever a value is mentioned, the property is, too, so there's less value in xreffing the value. Plus there are lots of values that have different meaning for different properties, and some might have the same name as a property, and so on. So I think I prefer leaving values alone.

zcorpan commented:
```
$ grep -Eo "&lt;[a-z-]+(>|&gt;)" source | sort | uniq -d
```

I couldn't find any more datatypes that weren't already xreffed.

This is great, btw. :+1:

zcorpan commented:
'vw' is mentioned in a note here https://html.spec.whatwg.org/#valid-source-size-list

cvrebert commented:
Glossed 'vw'.

zcorpan commented:
I think this looks pretty good now! Except I think we should skip the commits that gloss values. And maybe this needs rewrapping in places, I haven't checked that.

Should also squash into fewer commits; maybe one for properties, one for units, one for the rest?

cvrebert commented:
> And maybe this needs rewrapping in places, I haven't checked that.

To avoid needless labor, I haven't rewrapped at all yet. Will do the rewrapping once the changes are (modulo wrapping) fully approved.

cvrebert commented:
So, while I agree in general it doesn't seem strictly necessary to link every single value keyword (which is why I only linked 3 values rather than all of them), here's my rationale for the 3 values in question:

* 'ruby-base' (of the 'display' property)
  * Literally the only reference to CSSRUBY in the HTML spec is in relation to 'ruby-base'
  * `<p>For the purposes of the CSS ruby model, runs of children of <code>ruby</code> elements that are not <code>rt</code> or <code>rp</code> elements are expected to be wrapped in anonymous boxes whose <span>'display'</span> property has the value <span>'ruby-base'</span>. <ref spec=CSSRUBY></p>`
* 'inline-block' (of the 'display' property)
  * This is linked in relation to the term "'inline-block' box" that appears in the spec several times.
  * CSS defines the terms "block box" & "inline box", which are already glossed. Not glossing "'inline-block' box" in some way seems inconsistent.
  * None of the existing references to "'inline-block' box" clarify that 'display' is the relevant property.
* 'table-cell' (of the 'display' property)
  * Similar to 'inline-block', [the section that mentions 'table-cell'](https://html.spec.whatwg.org/#the-dialog-element:margin-box-10) doesn't mention the 'display' property

zcorpan commented:
OK that is reasonable.

cvrebert commented:
I'm happy to line-wrap and squash if no one has further comments.

zcorpan commented:
Let's wait for @domenic to check this

cvrebert commented:
@domenic LGTY?

domenic commented:
Reviewing now...

domenic commented:
- named colour is not linking correctly (presumably because it is only referenced as a plural)
- I'd like the border properties table to not be part of a bulleted list. I'd do a new paragraph, like "The CSS specification also defines the following border properties: [CSS] (... table goes here ...)"
- The border table needs an empty `<th>` as all the column headers are currently off by one.
- Units (em, ex, vh, in, and px) should not be quoted
- I think the CSS Color section with its three terms should move after the CSS specification section with its many terms. Having CSS itself be the first list-of-terms makes the most sense to me.

Otherwise LGTM... there's a lot of related messiness in these sections, but it's the same before and after your change (non-xlinked terms, missing spec ref from the intrinsic dimensions sentence, wrong titles for CSS specs (e.g. including "CSS3"). We'll get to those eventually, perhaps in a great xref-of-doom pull request.

domenic commented:
> Units (em, ex, vh, in, and px) should not be quoted

Looking at parts of CSS Values that are not headings, this appears to be wrong; sorry. Quoted is fine.

cvrebert commented:
@domenic 

> named colour is not linking correctly (presumably because it is only referenced as a plural)

(Note that that's a preexisting bug not caused by these changes.) Fixed.

> I'd like the border properties table to not be part of a bulleted list.

Done in 8793f0e.

> The border table needs an empty `<th>` as all the column headers are currently off by one.

Fixed in cfaf8f4.

> I think the CSS Color section with its three terms should move after the CSS specification section 

Done in 8835670.

domenic commented:
LGTM!

cvrebert commented:
Squashed & rebased.

cvrebert commented:
Oh, right, need to line-wrap...

cvrebert commented:
Okay, line-wrapped now.

domenic commented:
The lines seem to have been wrapped to something other than 100 characters. Would you mind wrapping them again? Looking at the combined diff it seems like there's a lot of violations of the 100-column limit.

Also, for the record my plan was to squash this into a single commit with the following message:

```
Update CSS cross references

- Fix link to "named colour" term
- Add direct spec references for all CSS properties, units, and types
- Add cross-references for the values of the 'display' property
- Update CSS2 link and verbiage

PR: https://github.com/whatwg/html/pull/782
```

Let me know if that misses anything.

cvrebert commented:
Augh, 740634d0f30a3b76e9da166ac2fa8835fcc073ab caused a nightmarish merge conflict.
Rebased again, squashed again, line-wrapped again.
(Note to self: old head was 4f18152)

domenic commented:
Yay! Merged as 50c6248c5250741d3e2184a2eb91a464522ceeaa.
